### DISCUSSION TITLE
Failure to write frames from Keyframe Selection

### DISCUSSION THREAD

**Discussion by saripsis** [alicevision/Meshroom#2209] [Q&A] [answered]
1) Made video
2) Drag and Drop into meshroom 2023.2.0
3) Computed the keyframe node
4) Wait
Result: Keyframes are selected, but can't be written to disk.
Tried with two different clips, 1 from DJI in .mp4, and one with a Cannon DSLR in .mov

Any help on telling me what I'm doing wrong would be appreciated.

>  [2023-09-26 14:44:48.687117] [0x00009724] [trace]   Embedded OCIO configuration file: 'C:\Meshroom-2023.2.0\aliceVision/share/aliceVision/config.ocio' found.
> WARNING:root:QQmlComponent: Created graphical object was not placed in the graphics scene.
>  - commandLine: aliceVision_keyframeSelection  --inputPaths "C:/tmp/DSC_0001.MOV" --brands "Nikon" --models "D5100" --sensorDbPath "C:\Meshroom-2023.2.0\aliceVision\share\aliceVision\cameraSensors.db" --useSmartSelection True --pxDisplacement 10.0 --minNbOutFrames 10 --maxNbOutFrames 2000 --rescaledWidthSharpness 720 --rescaledWidthFlow 720 --sharpnessWindowSize 200 --flowCellSize 90 --outputExtension none --exportScores False --exportFlowVisualisation False --flowVisualisationOnly False --skipSharpnessComputation False --skipSelection False --verboseLevel info --outputFolder "C:/tmp/MeshroomCache/KeyframeSelection/8773b23b447c2a970dd6bb397c7a6174712de566" --outputSfMDataKeyframes "C:/tmp/MeshroomCache/KeyframeSelection/8773b23b447c2a970dd6bb397c7a6174712de566/keyframes.sfm" --outputSfMDataFrames "C:/tmp/MeshroomCache/KeyframeSelection/8773b23b447c2a970dd6bb397c7a6174712de566/frames.sfm"
>  - logFile: C:/tmp/MeshroomCache/KeyframeSelection/8773b23b447c2a970dd6bb397c7a6174712de566/log
> ERROR:root:Error on node computation: Error on node "KeyframeSelection_1":
> Log:
> [2023-09-26 14:45:33.301771] [0x00007c98] [trace]   Embedded OCIO configuration file: 'C:\Meshroom-2023.2.0\aliceVision/share/aliceVision/config.ocio' found.
> Program called with the following parameters:
>  * brands =  = [Nikon]
>  * csvFilename = "scores.csv" (default)
>  * exportFlowVisualisation = 0
>  * exportScores = 0
>  * exportSelectedFrames = 0 (default)
>  * flowCellSize = 90
>  * flowVisualisationOnly = 0
>  * inputPaths =  = [C:/tmp/DSC_0001.MOV]
>  * maxCoresAvailable =  Unknown Type "unsigned int" (default)
>  * maxFrameStep =  Unknown Type "unsigned int" (default)
>  * maxMemoryAvailable = 18446744073709551615 (default)
>  * maxNbOutFrames =  Unknown Type "unsigned int"
>  * minFrameStep =  Unknown Type "unsigned int" (default)
>  * minNbOutFrames =  Unknown Type "unsigned int"
>  * mmFocals =  Unknown Type "class std::vector<float,class std::allocator<float> >" (default)
>  * models =  = [D5100]
>  * outputExtension = "none"
>  * outputFolder = "C:/tmp/MeshroomCache/KeyframeSelection/8773b23b447c2a970dd6bb397c7a6174712de566"
>  * outputSfMDataFrames = "C:/tmp/MeshroomCache/KeyframeSelection/8773b23b447c2a970dd6bb397c7a6174712de566/frames.sfm"
>  * outputSfMDataKeyframes = "C:/tmp/MeshroomCache/KeyframeSelection/8773b23b447c2a970dd6bb397c7a6174712de566/keyframes.sfm"
>  * pxDisplacement = 10
>  * renameKeyframes = 0 (default)
>  * rescaledWidthFlow = 720
>  * rescaledWidthSharpness = 720
>  * sensorDbPath = "C:\Meshroom-2023.2.0\aliceVision\share\aliceVision\cameraSensors.db"
>  * sharpnessWindowSize = 200
>  * skipSelection = 0
>  * skipSharpnessComputation = 0
>  * storageDataType =  Unknown Type "enum aliceVision::image::EStorageDataType" (default)
>  * useSmartSelection = 1
>  * verboseLevel = "info"
> 
> Hardware :
>         Detected core count : 20
>         OpenMP will use 20 cores
>         Detected available memory : 7294 Mo
> 
> [14:45:33.320785][info] Single camera
> [14:45:33.320785][info] Camera: C:/tmp/DSC_0001.MOV
>          - brand: Nikon
>          - model: D5100
>          - focal (mm): 0
> [14:45:33.503771][info] Finished processing frame 1/2178
> .....removed for readability
> [14:58:06.645252][info] Finished processing frame 2178/2178
> [14:58:06.657268][info] Selecting frame with ID 4
> .....removed for readability
> [14:58:06.660268][info] Selecting frame with ID 2169
> [14:58:06.660268][info] Selecting frame with ID 2175
> [14:58:06.660268][info] Finished selecting all the keyframes! 496/2178 frames have been selected.
> [14:58:06.761607][fatal] The keyframes selected from the input video have not been written on disk. The keyframes' SfMData file cannot be written.
> 
> WARNING:root:Retrying to obtain clipboard.
> WARNING:root:Retrying to obtain clipboard.

**Comment by cbentejac** [comment, marked as answer]
Hi @saripsis,

TL;DR: setting the "Ouptut Extension" parameter to anything but "none" should fix your problem. 

By default, the keyframes are not written on disk, only the SfMData files are (one containing the list of selected keyframes, one with the list of frames that were not selected). Those SfMData files use the paths of the input frames, which are already present on disk when you use the Keyframe Selection with sequences of images.

In the case of video inputs, the frames from the videos are not already written on disk, so the SfMData files cannot be written as such. The SfMData file containing the list of selected keyframes can be written if you select another option than "none" for the "Output Extension" parameter: the selected keyframes will be written on disk (with the selected file extension), and the SfMData will be correctly written (but the SfMData file containing the frames that were not selected will never be written for video inputs).

In the next version of Meshroom, the "Output Extension" parameter will be highlighted if it is set to "none" while the input is a video, specifically to prevent running into this issue.

**Reply by saripsis** [reply]
Thank you! Perfect answer.
